datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java

```diff
@@ -0,0 +1,219 @@
+/**
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * NOTE:
+ * SimpleStorageResource.java from spring-cloud-aws
+ * located by https://github.com/spring-cloud/spring-cloud-aws/blob/master/spring-cloud-aws-core/src/main/java/org/springframework/cloud/aws/core/io/s3/SimpleStorageResource.java
+ * is used and was modified according Adorsys project needs.
+ **/
+package de.adorsys.datasafe.storage.impl.s3;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.*;
+import com.amazonaws.util.BinaryUtils;
+import de.adorsys.datasafe.types.api.resource.ResourceLocation;
+import de.adorsys.datasafe.types.api.utils.Log;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.Executors;
+
+@Slf4j
+public class MultipartUploadS3StorageOutputStream extends OutputStream {
+
+    private String bucketName;
+
+    private String objectName;
+
+    private AmazonS3 amazonS3;
+
+    // The minimum contentSize for a multi part request is 5 MB, hence the buffer contentSize of 5 MB
+    private static final int BUFFER_SIZE = 1024 * 1024 * 5;
+
+    private final Object monitor = new Object();
+
+    private final CompletionService<UploadPartResult> completionService;
+
+    private ByteArrayOutputStream currentOutputStream = new ByteArrayOutputStream(BUFFER_SIZE);
+
+    private InitiateMultipartUploadResult multiPartUploadResult;
+
+    private int partCounter = 1;
+
+    public MultipartUploadS3StorageOutputStream(String bucketName, ResourceLocation resource, AmazonS3 amazonS3) {
+        this.bucketName = bucketName;
+        this.objectName = resource.location().getPath().replaceFirst("^/", "");;
```

Comment:
This code is a copy paste of S3StorageService. Further it is not a good idea to choose the delimiter hard coded. The delimiter can be any character.

Reply:
I think we have to replace it everywhere, not just here. Let's create new issue for that